File: tests/bench/crosslang/c/todo.c
/* Порт tests/bench/programs/TodoBench.eat 1:1 — ядро RESTful TODO-list:
 * bounded-пул из 64 слотов под потоком create/toggle/serialize/remove.
 * title зависит от loop-carried acc, поэтому сериализацию не сфолдить.
 * Вывод обязан совпасть байт-в-байт с EATLang-оригиналом при том же
 * REPEAT. */
#include <stdint.h>
#include <stdio.h>

static const uint32_t REPEAT = 1;

#define CAP 64u
#define TLEN 16u

struct Store {
    uint32_t id[64];
    uint8_t title[1024];
    uint8_t done[64];
    uint8_t used[64];
    uint32_t next_id;
};

static uint32_t st_create(struct Store *s, uint32_t k, uint32_t seed) {
    uint32_t slot = k % CAP;
    uint32_t base = slot * TLEN;
    uint32_t id = s->next_id;
    s->id[slot] = id;
    for (uint32_t j = 0; j < 16u; j++) {
        s->title[base + j] = (uint8_t)(65u + (seed + k + j * 7u) % 26u);
    }
    s->done[slot] = (uint8_t)(k % 2u);
    s->used[slot] = 1;
    s->next_id = s->next_id + 1u;
    return id;
}

static void st_toggle(struct Store *s, uint32_t idx) {
    if (s->done[idx] == 0) {
        s->done[idx] = 1;
    } else {
        s->done[idx] = 0;
    }
}

static void st_remove(struct Store *s, uint32_t idx) { s->used[idx] = 0; }

static uint32_t st_serialize(const struct Store *s) {
    uint32_t sum = 0;
    for (uint32_t slot = 0; slot < 64u; slot++) {
        if (s->used[slot] == 1) {
            uint32_t m = s->id[slot];
            for (uint32_t d = 0; d < 10u; d++) {
                if (m == 0) {
                    break;
                }
                sum = (sum * 31u + m % 10u) % 65536u;
                m = m / 10u;
            }
            uint32_t base = slot * TLEN;
            for (uint32_t j = 0; j < 16u; j++) {
                sum = (sum * 31u + s->title[base + j]) % 65536u;
            }
            sum = (sum * 31u + s->done[slot]) % 65536u;
        }
    }
    return sum;
}

int main(void) {
    struct Store s = {0};
    s.next_id = 1;
    uint32_t acc = 0;
    for (uint32_t r = 0; r < REPEAT; r++) {
        for (uint32_t k = 0; k < 2048u; k++) {
            uint32_t seed = acc;
            uint32_t id = st_create(&s, k, seed);
            acc = (acc * 31u + id) % 65536u;
            uint32_t t = (k * 3u) % CAP;
            st_toggle(&s, t);
            if (k % 8u == 0) {
                acc = (acc * 31u + st_serialize(&s)) % 65536u;
            }
            if (k % 16u == 0) {
                uint32_t rr = (k * 5u) % CAP;
                st_remove(&s, rr);
            }
        }
    }
    printf("checksum %u\n", acc);
    return 0;
}

File: tests/bench/crosslang/go/todo.go
// Порт tests/bench/programs/TodoBench.eat 1:1 — ядро RESTful TODO-list:
// bounded-пул из 64 слотов под потоком create/toggle/serialize/remove.
// Вывод обязан совпасть байт-в-байт с EATLang-оригиналом при том же
// REPEAT.
package main

import "fmt"

const REPEAT uint32 = 1
const CAP uint32 = 64
const TLEN uint32 = 16

type Store struct {
	id     [64]uint32
	title  [1024]uint8
	done   [64]uint8
	used   [64]uint8
	nextID uint32
}

func (s *Store) create(k, seed uint32) uint32 {
	slot := k % CAP
	base := slot * TLEN
	id := s.nextID
	s.id[slot] = id
	for j := uint32(0); j < 16; j++ {
		s.title[base+j] = uint8(65 + (seed+k+j*7)%26)
	}
	s.done[slot] = uint8(k % 2)
	s.used[slot] = 1
	s.nextID++
	return id
}

func (s *Store) toggle(idx uint32) {
	if s.done[idx] == 0 {
		s.done[idx] = 1
	} else {
		s.done[idx] = 0
	}
}

func (s *Store) remove(idx uint32) { s.used[idx] = 0 }

func (s *Store) serialize() uint32 {
	var sum uint32 = 0
	for slot := uint32(0); slot < 64; slot++ {
		if s.used[slot] == 1 {
			m := s.id[slot]
			for d := 0; d < 10; d++ {
				if m == 0 {
					break
				}
				sum = (sum*31 + m%10) % 65536
				m = m / 10
			}
			base := slot * TLEN
			for j := uint32(0); j < 16; j++ {
				sum = (sum*31 + uint32(s.title[base+j])) % 65536
			}
			sum = (sum*31 + uint32(s.done[slot])) % 65536
		}
	}
	return sum
}

func main() {
	s := Store{nextID: 1}
	var acc uint32 = 0
	for r := uint32(0); r < REPEAT; r++ {
		for k := uint32(0); k < 2048; k++ {
			seed := acc
			id := s.create(k, seed)
			acc = (acc*31 + id) % 65536
			t := (k * 3) % CAP
			s.toggle(t)
			if k%8 == 0 {
				acc = (acc*31 + s.serialize()) % 65536
			}
			if k%16 == 0 {
				rr := (k * 5) % CAP
				s.remove(rr)
			}
		}
	}
	fmt.Printf("checksum %d\n", acc)
}
